lib/vrl/stdlib/src/reverse_dns.rs

```diff
@@ -0,0 +1,92 @@
+use dns_lookup::lookup_addr;
+use std::net::IpAddr;
+
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ReverseDns;
+
+impl Function for ReverseDns {
+    fn identifier(&self) -> &'static str {
+        "reverse_dns"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::BYTES,
+            required: true,
+        }]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "Example",
+            source: r#"reverse_dns!("127.0.0.1")"#,
+            result: Ok("localhost"),
+        }]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::Compiler,
+        _info: &FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+
+        Ok(Box::new(ReverseDnsFn { value }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ReverseDnsFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for ReverseDnsFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let ip: IpAddr = self
+            .value
+            .resolve(ctx)?
+            .try_bytes_utf8_lossy()?
+            .parse()
+            .map_err(|err| format!("unable to parse IP address: {}", err))?;
+
+        let host =
+            lookup_addr(&ip).map_err(|err| format!("unable to perform a lookup : {}", err))?;
+
+        Ok(host.into())
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value.type_def(state).fallible().bytes()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    test_function![
+        reverse_dns => ReverseDns;
+
+        invalid_ip {
+            args: func_args![value: value!("999.999.999.999")],
+            want: Err("unable to parse IP address: invalid IP address syntax"),
+            tdef: TypeDef::new().fallible().bytes(),
+        }
+
+        localhost {
```

Comment:
Is it worth putting in a test for `::1`?

Reply:
I ended up dropping the localhost tests since Windows doesn't return `localhost`. I did add an IPv6 test for `dns.google` though in 6fe11b700